### PR TITLE
Whitelist zeep module (SOAP)

### PR DESCRIPTION
**Pull request by nngu6036**

### CODE REVIEW

File: collective/trustedimports/suds.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from collective.trustedimports.util import whitelist_module
+
+whitelist_module('suds')
```
Comment (djay) [inline collective/trustedimports/suds.py]: Looking at this documentation this accepts urls which makes it a security risk.
I think what we need to do is have a env variable called SAFEIMPORTS_URL_BLACKLIST with a list of urls that should not be allowed. And we patch any function that accesses the internet to check this list first and throw an exception.

 https://bitbucket.org/jurko/suds/wiki/Original%20Documentation. Also you will likely need to whitelist more things that this to make it work.

**Review by djay — CHANGES_REQUESTED** [review]

File: collective/trustedimports/suds.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from suds import MethodNotFound
+
+
+class SafeClient(_Client):
+    """ SafeClient
+    """
+
+    def __init__(self, url, **kwargs):
+
+        blacklist = os.getenv('SAFEIMPORTS_URL_BLACKLIST ', '')
```
Comment (djay) [inline collective/trustedimports/suds.py]: needs to support wildcards.
probably should be in the utils for other libraries that access urls to use

Needs to only do the blacklist when being called via a restrictedpython same as https://github.com/collective/collective.trustedimports/blob/master/collective/trustedimports/safezipfile.py#L20
Maybe move that check function into utils

File: collective/trustedimports/suds.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        if url in blacklist:
+            raise Unauthorized("Not supported by SafeClient in this context")
```
Comment (djay) [inline collective/trustedimports/suds.py]: wrong kind of error. it shouldn't lead to a login. It should the the same error as when you do something else not allowed in restrictedpython like a.b = 1.

**Review by djay — CHANGES_REQUESTED** [review]

File: collective/trustedimports/suds.py (hunk 2 above)
Comment (djay) [reply collective/trustedimports/suds.py]: @nngu6036 I've written you a helper function in utils to make this much easier. You just need to check the rest of the functions to make sure they are safe and write the tests

Comment (nngu6036) [reply collective/trustedimports/suds.py]: @djay , I add the test

File: collective/trustedimports/suds.py (hunk 1 above)
Comment (nngu6036) [reply collective/trustedimports/suds.py]: @djay , I implement wildcard in blacklist, using fmatch, the blacklist use semicolon as delimiter

File: collective/trustedimports/util.py [hunk 4]
```diff
@@ -41,8 +42,11 @@ def restricted_python_call():
 def is_url_allowed(url=None, uri=None, link=None):
     url = [name for name in [url,uri,link] if name is not None]
     blacklist = os.getenv('SAFEIMPORTS_URL_BLACKLIST ', '')
-    # TODO: better test for wildcard domain names
-    return url not in blacklist
+    # Use semicolin delimiter for multiple url in blacklist
+    for pattern in re.split(';',blacklist):
```
Comment (djay) [inline collective/trustedimports/util.py]: handle extra whitespace.

File: collective/trustedimports/suds.rst [hunk 5]
```diff
@@ -0,0 +1,16 @@
+Zip
```
Comment (djay) [inline collective/trustedimports/suds.rst]: ?

File: collective/trustedimports/suds.rst [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+When used in RestrictedPython we can still use Suds
+
+>>> teval("from suds.client import Client; from suds import WebFault,MethodNotFound;return Client('')")
+Traceback (most recent call last):
+...
+ValueError: unknown url type:
+
+>>> from suds.client import Client; from suds import WebFault,MethodNotFound;
+>>> Client('')
```
Comment (djay) [inline collective/trustedimports/suds.rst]: you aren't testing what you are importing

File: collective/trustedimports/suds.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
+Zip
+===
+
+When used in RestrictedPython we can still use Suds
+
+>>> teval("from suds.client import Client; from suds import WebFault,MethodNotFound;return Client('')")
```
Comment (djay) [inline collective/trustedimports/suds.rst]: How are you testing the result of a client call and subaccess to it?
e..g https://webkul.com/blog/python-suds-client/

**Review by djay — CHANGES_REQUESTED** [review]

File: collective/trustedimports/util.py (hunk 4 above)
Comment (nngu6036) [reply collective/trustedimports/util.py]: @djay  I trim the pattern

File: collective/trustedimports/tests.py [hunk 8]
```diff
@@ -34,7 +35,7 @@ def setUp(test=None):
     MSI('sets').declarePublic('Set')
     newSecurityManager(None, None)
     xmlconfig.file('configure.zcml', collective.trustedimports, ) #context=configurationContext)
-
+    os.environ["SAFEIMPORTS_URL_BLACKLIST"] = "https://www.w3schools.com/Xml/tempconvert.asmx?WSDL;http://www.dneonline.com/*.asmx?WSDL"
```
Comment (djay) [inline collective/trustedimports/tests.py]: this is for a specific test so it should go there. it also needs to be removed afterwards

File: collective/trustedimports/soap.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+>>> teval("from zeep import Client;return Client('http://webservices.oorsprong.org/websamples.countryinfo/CountryInfoService.wso?WSDL')")
+<zeep.client.Client ... at ...>
+
+>>> teval("from zeep.exceptions import Error;raise Error('Test error')")
+Traceback (most recent call last):
+...
+Error: Test error
+
+
+We cann't set transport option for WSDL service
```
Comment (djay) [inline collective/trustedimports/soap.rst]: The code below doesn't show setting a transport

File: collective/trustedimports/soap.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+>>> teval("from zeep import Client;client = Client('http://webservices.oorsprong.org/websamples.countryinfo/CountryInfoService.wso?WSDL', settings = {'force_https':True})")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl,settings' have values not supported in a restricted python call
```
Comment (djay) [inline collective/trustedimports/soap.rst]: why can't you use these settings? You should be able to

File: collective/trustedimports/soap.rst [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+>>> teval("from zeep import Client;client = Client('http://webservices.oorsprong.org/websamples.countryinfo/CountryInfoService.wso?WSDL', settings = {'extra_http_headers':''})")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl,settings' have values not supported in a restricted python call
+
+We cann't open connection to URL in blacklist
```
Comment (djay) [inline collective/trustedimports/soap.rst]: set the blacklist env here. make sure you remove it after

File: collective/trustedimports/soap.rst [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+>>> teval("from zeep import Client;return Client('http://www.dneonline.com/calculator.asmx?WSDL')")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl' have values not supported in a restricted python call
+
+We cann't call method 'bind'
+>>> teval("from zeep import Client;return Client('https://www.w3schools.com/Xml/tempconvert.asmx?WSDL');client.bind()")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl' have values not supported in a restricted python call
```
Comment (djay) [inline collective/trustedimports/soap.rst]: That error doesn't have anything to do with bind so something is wrong here

File: collective/trustedimports/soap.rst [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+>>> teval("from zeep import Client;return Client('https://www.w3schools.com/Xml/tempconvert.asmx?WSDL');client.create_service()")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl' have values not supported in a restricted python call
+
+We can call WSDL method as usual
+>>> teval("from zeep import Client; client = Client('http://webservices.oorsprong.org/websamples.countryinfo/CountryInfoService.wso?WSDL');result=client.service.CapitalCity('NL');assert result == 'Amsterdam'")
+
+
+We can still use it when not in a PythonScript
```
Comment (djay) [inline collective/trustedimports/soap.rst]: You need to show you can use something restricted not something that works in both

File: collective/trustedimports/soap.rst [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+...
+ValueError: Argument(s) 'wsdl' have values not supported in a restricted python call
+
+We cann't call method 'bind'
+>>> teval("from zeep import Client;return Client('https://www.w3schools.com/Xml/tempconvert.asmx?WSDL');client.bind()")
+Traceback (most recent call last):
+...
+ValueError: Argument(s) 'wsdl' have values not supported in a restricted python call
+
+We cann't call method 'create_service'
```
Comment (djay) [inline collective/trustedimports/soap.rst]: same as above

**Review by djay — CHANGES_REQUESTED** [review]

File: collective/trustedimports/soap.rst (hunk 9 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , add test for transport

File: collective/trustedimports/soap.rst (hunk 10 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , I think that these options may be security risks

File: collective/trustedimports/soap.rst (hunk 11 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , I move the env code into doctest

File: collective/trustedimports/soap.rst (hunk 12 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , fix the test with disallow function

File: collective/trustedimports/soap.rst (hunk 14 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , fix the test with disallow fuction

File: collective/trustedimports/soap.rst (hunk 13 above)
Comment (nngu6036) [reply collective/trustedimports/soap.rst]: @djay , I add the test to use bind function in normal case

**Pull request merged**